File: day_6/puzzle_2/solve.py
import os
import sys
from typing import List

sys.path.append(
    os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
)

from benchmark.benchmark import benchmark


def main():
    with open("./day_6/input.txt", "r") as file:
        packet_stream = file.read()

        start_of_packet_marker = -1

        for i in range(len(packet_stream)):
            packet = packet_stream[i : i + 14]
            checked: List[str] = []

            for value in packet:
                if value in checked:
                    break
                checked.append(value)

            if len(checked) == 14:
                start_of_packet_marker = i + 14
                break

        return start_of_packet_marker


if __name__ == "__main__":
    benchmark(main, "day_6_puzzle_2")

File: day_6/puzzle_2/solve.rb
# frozen_string_literal: true

require_relative '../../benchmark/benchmark'

Benchmark.run('day_6_puzzle_2') do
  File.open('./day_6/input.txt', 'r') do |file|
    packet_stream = file.read

    start_of_packet_marker = -1

    packet_stream.length.times do |i|
      packet  = packet_stream[i..i + 13]
      checked = []

      packet.each_char do |value|
        break if checked.include?(value)

        checked.push(value)
      end

      if checked.length == 14
        start_of_packet_marker = i + 14
        break
      end
    end

    start_of_packet_marker
  end
end
